### ISSUE TITLE
QScript needs to merge `LeftShift` containing `Typecheck`

### ISSUE THREAD

**Issue by alissapajer** [quasar-analytics/quasar#2133]
In the query
```
select * from nested_foo where (foo{*} = 5 or foo[*] = 6)
```
we need to merge the two sides of the `or` in qscript. Currently this fails to merge in qscript and creates an unnecessary theta join.

Here is the relevance `LogicalPlan`:
```
      ╰─ Invoke(Or)
         ├─ Invoke(Eq)
         │  ├─ Invoke(FlattenMap)
         │  │  ╰─ Let('__tmp2)
         │  │     ├─ ObjectProject(Free('__tmp0){"foo"})
         │  │     ╰─ Typecheck(Obj(Map[], Some(Top)))
         │  │        ├─ Free('__tmp2)
         │  │        ├─ Free('__tmp2)
         │  │        ╰─ Constant(Obj(Map( -> NA)))
         │  ╰─ Constant(Int(5))
         ╰─ Invoke(Eq)
            ├─ Invoke(FlattenArray)
            │  ╰─ Let('__tmp3)
            │     ├─ ObjectProject(Free('__tmp0){"foo"})
            │     ╰─ Typecheck(FlexArr(0, None, Top))
            │        ├─ Free('__tmp3)
            │        ├─ Free('__tmp3)
            │        ╰─ Constant(Arr(List(NA)))
            ╰─ Constant(Int(6))
```
This should enable the integration test `multitypeFlatten.test`. It is also possible that this work will enable a handful of other flatten-related integration tests.

**Comment by alissapajer** [comment]
Had a discussion about this with @djspiewak. It turns out that the problem here is a fundamental limitation to the current qscript merge capabilities.

Currently, as we merge sources in qscript, we require that the left and right sides each remain accessible through a `FreeMap`. In this case we need each side to remain accessible through a `LeftShift`.

Ideally the qscript resulting from this query would look approximately like:
```
Filter(nestedFoo,
  Or(
    Eq(LeftShift((), Guard(ProjectObject(foo), Object), RightSide), 5), 
    Eq(LeftShift((), Guard(ProjectObject(foo), Array), RightSide), 6)))
```
But this is currently invalid qscript because only `FreeMap` are allowed in the operation of a `Filter`, yet this proposed qscript has `LeftShift` in that position.

You may wonder why the common source in the merge cannot be a `LeftShift`? If we were to do that, we'd need a special modified `LeftShift` that can express two different guards for different parts of the query that follows. This would be special-casing `LeftShift` in a way that doesn't generalize. Similarly you may wonder why we cannot have something like `LeftShift(_, Or(Guard(Obj), Guard(Arr)), _, _)` or `LeftShift(_, Guard(Obj \/ Arr), _, _)`: these again don't let us tie the `Guards` back to their corresponding parts in the query. 

Making this change in qscript also affects all backends, because this fix is essentially expanding `FreeMap` into something like `Coproduct[FreeMap, LeftShift]`. That may not be the exact type we end up with, but that is the general idea. As such, all backends will need to handle this new type.

Also note that the `multitypeFlatten.test` should be producing _correct_ qscript without this fix (even though that qscript contains a join). So this ticket is only to modify the qscript so that it doesn't produce an unnecessary join, and not to enable the test in various backends.

**Comment by alissapajer** [comment]
Also note that #2110 exemplifies essentially the same problem, so we need to fix this issue generally enough that #2110 will also be fixable in the same way.

**Comment by sellout** [comment]
I think the QScript should look roughly like this:

```
Union(
  ShiftedRead("/nestedFoo"),
  Filter(
    LeftShift(HoleF, Guard(ProjectField(HoleF, "foo"), Object, …), RightSide),
    Eq(HoleF, 5)),
  Filter(
    LeftShift(HoleF, Guard(ProjectField(HoleF, "foo"), Array, …), RightSide),
    Eq(HoleF, 6)))
```

**Comment by djspiewak** [comment]
You made the Or directional with that encoding. Any foo which is an array and not an object will fail.

**Comment by djspiewak** [comment]
Oh wait union. Never mind. That works, though it only solves this special case, and not the more general one in the other ticket.

**Comment by sellout** [comment]
I think this and #2110 are distinct problems – I mean, they’re similar in that we have to deal with extracting `LeftShift` from the mapping operations in the monosorted LP structure, but this one applies two different shifts to a single src, while the other applies LeftShifts to two joined srcs.

**Comment by alissapajer** [comment]
@sellout hm, yeah that's an important point. Though, I think the same solution still might apply in the case of `LeftShift`s applied to two joined sources, and here's why. 

The join condition comes to qscript from logical plan containing `lp.JoinSideName`, which get translated into the qscript `MapFunc` called `JoinSideName`. So, we actually just have `MapFunc`s pointing to each side of the join, and `MapFunc`s are mergeable. So if we are merging a `LeftShift` of `JoinSideName(__leftside)` and a `LeftShift` of `JoinSideName(__rightside)` then those should still merge (and not create an intermediate theta join) because they are just pointers (mappings) to the sides and not the sides themselves. We don't replace them with their actual sides (`LeftSide` and `RightSide`) until we hit the `lp.Join` case itself. There are effectively are no sources in the join condition, which is why we match on `case Some(Map(_, mf)) if mf.count ≟ 0 =>` to obtain what becomes the join condition in qscript.

**Comment by alissapajer** [comment]
Though, as I think about this, the sources we are merging still have to "line up" in order to pass through the merge algorithm, and there is no guaranteed symmetry across a join condition of course. So things won't actually merge in as many cases as I was originally thinking. 

But I _would_ still like to avoid creating a theta join in the join condition at all, because it's so nice for qscript when the join condition comes handed to us nicely. And it's a real mess if we have to dissect theta joins in qscript. So, maybe we can make a change in logical paln (at least for the two joined sources case) where we break out shifts inside join conditions so we're prepared to handle them in qscript.

**Comment by alissapajer** [comment]
But dealing with this in logical plan might be just as messy. Because it's not like we're just shifting once in a join condition. We can have a condition that is `foo[*]{*}.bar[2].baz = ack[1]{*}.arg[*][*]` where we need to shift _and_ perform mapping operations in a dovetailed sort of way. So, hrm. Not sure what's best here yet...

**Comment by alissapajer** [comment]
But if `LeftShift` really is "just another mapping operation" (in the sense of having a sort of `Coproduct[MapFunc, LeftShift]`)  then actually I think the merging would be fine, even if the sides don't "line up" because it would be perfectly fine if a `LeftShift` on one side lines up with an `ObjectProject` on the other because those are both mapping things that let us continue through the merge. So, now I'm back to thinking these are the same issue. We'll have to see what I think in the morning.

**Comment by sellout** [comment]
But it’s not just another mapping operation, it’s flatMap-ish. And not pulling it out means the onus is on the connector authors to identify and handle that. It breaks our mapping / dimensional distinction.

**Comment by alissapajer** [comment]
What I was trying to say is that maybe we can afford to treat shifting similarly to how we treat mapping when we are merging sources in qscript. But yes that would force connectors to handle a shifting operation in a place where they would previously have had to handle only mapping operations. I guess the question is, how bad would that be? 

I am completely on board with making things for backend authors as easy as possible, so if this needs to be done in qscript (or sooner) for that reason, I'm all for it. But I was just thinking that backends already implement `LeftShift` and `FreeMap` handling as part of `QSciptCore`, so maybe it wouldn't be that hard to have either one of these allowed in certain places? But maybe this really makes things hard for backends? A concrete example of what would get harder for a particular backend would be helpful.

**Comment by sellout** [comment]
Sure, so in Mongo, one of the simplifications we’ve been able to make with QScript is that the handling of Selectors, Expressions, and JavaScript only has to handle MapFuncs, which is good, because these are the only operations they support[†]. When we were compiling Mongo from LP, we generated an `ExprBuilder` (akin to `qscript.Map`), or some other builder, at each step, where the builders represent some abstraction over the aggregation pipeline. Then we had implemented various operations that paralleled `qscript.Coalesce`, etc. in order to turn `ExprBuilder(ExprBuilder(src, $field("foo")), $length($$ROOT))` into `ExprBuilder(src, $length($field("foo")))`.

Moving to QScript meant that we could get rid of most of that (and I keep eliminating more of it).

LeftShift is handled by either $unwind (in the aggregation pipline, if we’re shifting an array) or by a “flatMap” operation in mapReduce (which can handle any shifting). mapReduce actually does mix map/flatmap in one operation, but aggregation pipeline doesn’t. Since we often don’t know whether we’re going to want to use $unwind or mapReduce until we’ve figured out exactly which pieces only exist in JS, the Mongo connector abstracts over this with a `FlatteningBuilder`.

So, mixing LeftShift into MapFuncs means that we need to go back to having the Expression handler (and Selector and JS handlers) return builders and having ways to simplify those builders, so we end up reintroducing the operations that QScript let us eliminate. And while I do think that QScript would still be beneficial in eliminating a lot of the guesswork in the old Mongo connector, there would need to be a bunch of code reintroduced to handle it.

[†]: JS is obviously more powerful, but it’s still easier to handle its use for reductions, etc. separately from its use in mapping (not least because these have to be done in separate functions in mapReduce).
